subprojects/core/src/test/groovy/org/gradle/api/internal/provider/DefaultListPropertyTest.groovy

```diff
@@ -123,4 +123,111 @@ class DefaultListPropertyTest extends PropertySpec<List<String>> {
         }
         0 * _
     }
+
+    def "can add value to the provider"() {
+        expect:
+        def property = new DefaultListProperty<String>(String)
+        property.set(["abc"])
+
+        property.present
+        def v = property.get()
+        v instanceof ImmutableList
+        v == ["abc"]
+
+        property.add("123")
+
+        def v2 = property.get()
+        v2 instanceof ImmutableList
+        v2 == ["abc", "123"]
+
+        property.add(Providers.of("456"))
+
+        def v3 = property.get()
+        v3 instanceof ImmutableList
+        v3 == ["abc", "123", "456"]
+
+        property.addAll(Providers.of(["789"]))
+
+        def v4 = property.get()
+        v4 instanceof ImmutableList
+        v4 == ["abc", "123", "456", "789"]
+    }
+
+    def "can add provider to an empty provider"() {
+        def provider = Stub(ProviderInternal)
+        provider.get() >>> ["123", "abc"]
+
+        expect:
+        def property = new DefaultListProperty<String>(String)
+        property.add(provider)
+
+        def v = property.get()
+        v instanceof ImmutableList
+        v == ["123"]
+
+        def v2 = property.get()
+        v2 instanceof ImmutableList
+        v2 == ["abc"]
+    }
+
+    def "can add provider of a collection to an empty provider"() {
+        def provider = Stub(ProviderInternal)
+        provider.get() >>> [["123"], ["abc"]]
+
+        expect:
+        def property = new DefaultListProperty<String>(String)
+        property.addAll(provider)
+
+        def v = property.get()
+        v instanceof ImmutableList
+        v == ["123"]
+
+        def v2 = property.get()
+        v2 instanceof ImmutableList
+        v2 == ["abc"]
+    }
+
+    def "can add provider of a collection to the provider"() {
+        def provider = Stub(ProviderInternal)
+        provider.get() >>> [["123"], ["abc"]]
+
+        expect:
+        def property = new DefaultListProperty<String>(String)
+        property.set(["aaa"])
+        property.addAll(provider)
+
+        def v = property.get()
+        v instanceof ImmutableList
+        v == ["aaa", "123"]
+
+        def v2 = property.get()
+        v2 instanceof ImmutableList
+        v2 == ["aaa", "abc"]
+    }
+
+    def "can set to null value to remove value after adding values"() {
+        given:
+        def property = property()
+        property.add("abc")
+        property.set(null)
```

Comment:
Similarly...  what would this mean?

```
property.set(null)
property.addAll(...)
```

Reply:
What is in `addAll` should be the result. It's pretty much test https://github.com/gradle/gradle/pull/3108/files#diff-36f214e87f3139ad38f7ba4cb9b19bc1R173

Reply:
`property.set(null)` does something a little different:
https://github.com/gradle/gradle/blob/bf2ebb95d6407f327d79d34a666424877d6473a7/subprojects/core/src/main/java/org/gradle/api/internal/provider/DefaultListProperty.java#L82

when you query a `Providers.notDefined()` Provider, I think it explodes.

Reply:
`PropertyState.set(null)` currently means "discard the value of this property"

Reply:
We will keep this behavior as is. @big-guy and I discussed adding a `clear` method to remove all element, however, we settled on using `PropertyState.set([])` for this behavior.